DataDog/dd-trace-dotnet
# [Profiler] Wrap socket syscalls to prevent application errors

## Summary of changes

Wrap some socket functions/syscall from libc to avoid causing applications errors.

## Reason for change

We have customer that uses OpenLDAP library to query his server. This library uses glibc functions (like `select`, `poll`...) (wrapper on system calls) that are not restartable by the kernel. (for more info `man 7 signal`). The OpenLDAP client library implements  the usual restart mechanism but leaves the choice to restart or not to the customer (there is setting specifying if it should restart or not) and this setting is `do not restart` by default.
```
		do {
			fd.revents = 0;
			rc = poll( &fd, 1, timeout );
		} while( rc == AC_SOCKET_ERROR && errno == EINTR &&
			LDAP_BOOL_GET(&ld->ld_options, LDAP_BOOL_RESTART ));
```
[Example with `poll`](https://github.com/openldap/openldap/blob/2738a32de3a324fc56effd44c2fedaff1a359011/libraries/libldap/os-local.c#L253C3-L257C56)

So when the profiler interrupts the execution of one of these functions, the kernel does not restart them, the function returns an error and the application layer consider it as error and fails by throwing an exception.

It's not all, this issue helped us uncovered another problem: if `recv` (or other receiving/sending operations) syscall on a blocking socket has a timeout, and the thread calling it keeps on being interrupted by the profiler, the timeout will never be honored and the call blocks forever.

This change implements a restart mechanism (based on the `EINTR` + returned error) but checks if the interruption comes from the profiler and ensure that the profiler is not too much intrusive with socket operations with timeout.

## Implementation details

- Get the syscalls for OpenLdap issue + sending/receiving functions from `man 7 signal`
- Implement the restart mechanism.
- Add a class which is responsible to interact with the wrapper and the profiler libraries: before calling the `real` function, calls into the profiler library to register a memory area. When the call is done check if the value of this memory area is still `0`.
- When the profiler interrupts the thread execution, it will set a non-zero value to the memory area registered previously.

## Test coverage
- Add a new integration test with an OpenLDAP server.
- Add a test to timeout test to ensure the timeout on socket operations is still honored.

## Other details
